File: clojure/src/net_perspective/peer/scheduler.clj
(ns net-perspective.peer.scheduler
  "10-minute batch scheduler with cancel-replace semantics."
  (:require [net-perspective.peer.batch :as batch]))

(def batch-interval-ms (* 10 60 1000))

(defn run-scheduler!
  "Starts the batch scheduler. Runs the batch immediately, then every
   10 minutes. Cancel-replace: if the previous batch is still running when
   the next tick fires, it is cancelled before the new one starts.

   Returns a zero-argument stop-fn. Calling stop-fn cancels any running
   batch and stops the scheduler loop."
  [server]
  (let [current-future (atom nil)
        stop?          (atom false)

        run-batch! (fn []
                     (let [prev @current-future]
                       (when (and prev (not (future-done? prev)))
                         (future-cancel prev)))
                     (reset! current-future
                             (future
                               (try (batch/run-batch! server)
                                    (catch Exception e
                                      (println (str "scheduler batch error: " (.getMessage e))))))))

        scheduler (future
                    (run-batch!)
                    (loop []
                      (Thread/sleep batch-interval-ms)
                      (when-not @stop?
                        (run-batch!)
                        (recur))))]

    (fn stop []
      (reset! stop? true)
      (future-cancel scheduler)
      (let [curr @current-future]
        (when (and curr (not (future-done? curr)))
          (future-cancel curr))))))

File: go/pkg/peer/scheduler.go
package peer

import (
	"context"
	"log"
	"time"
)

const batchInterval = 10 * time.Minute

// RunScheduler runs the batch update every 10 minutes.
// It cancels and replaces any in-progress run before starting the next.
// Call with a parent context; returns when the parent is cancelled.
func (s *Server) RunScheduler(ctx context.Context) {
	var (
		runCancel context.CancelFunc
		done      = make(chan struct{}, 1)
	)
	runBatch := func() {
		if runCancel != nil {
			runCancel() // cancel prior run
			<-done      // wait for it to exit
		}
		var runCtx context.Context
		runCtx, runCancel = context.WithCancel(ctx)
		go func() {
			defer func() { done <- struct{}{} }()
			if err := s.RunBatch(runCtx); err != nil && runCtx.Err() == nil {
				log.Printf("batch run error: %v", err)
			}
		}()
	}

	// Run immediately on startup, then on each tick.
	runBatch()
	ticker := time.NewTicker(batchInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			if runCancel != nil {
				runCancel()
				<-done
			}
			return
		case <-ticker.C:
			runBatch()
		}
	}
}
